parsers/Parser.js

```diff
@@ -7,31 +9,97 @@ module.exports = class Parser {
         this.document = document;
     }
 
-    parse() {
-        return [];
+    getRules() {
+        const rules = this.rules;
+
+        return rules && rules.length ? this.flagDuplicates(rules) : [];
     }
 
-    getRules() {
-        let rules = this.parse();
+    get rules() {
+        // This default implementation fits for for YAML and JSON parsers.
+        // JS & Package parsers override it.
+        return this.getAllRules(this.ast);
+    }
 
-        if (!rules || rules.length === 0) {
-            return rules;
-        }
+    get ast() {
+        throw new Error('ast getter must be implemented in the child class.');
+    }
 
-        return this.flagDuplicates(rules);
+    getNodeKey(node) {
+        // This default implementation fits for YAML, JSON, and Package parsers.
+        // JS parser overrides it.
+        return node.key.value;
+    }
+
+    /** Gets properties of an object-like AST node. */
+    getProps(node) {
+        throw new Error('getProps method must be implemented in the child class.');
+    }
+
+    /** Gets elements of an array-like AST node. */
+    getItems(node) {
+        throw new Error('getItems method must be implemented in the child class.');
+    }
+
+    /** Flatmaps `callback` over `node`'s child property (or children properties) named `key`. */
+    mapChildren(node, key, callback) {
+        return this.getProps(node)
+            .filter(prop => this.getNodeKey(prop) === key)
+            .flatMap(prop => callback(prop.value));
```

Comment:
Using `flatMap` requires users of this extension to have a version of Node that is fairly new (11).  Not everyone can update right away and we should avoid forcing them to do so.